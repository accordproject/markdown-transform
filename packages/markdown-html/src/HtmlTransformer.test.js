/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// @ts-nocheck
/* eslint-disable no-undef */
'use strict';

const fs = require('fs');
const CiceroMarkTransformer = require('@accordproject/markdown-cicero').CiceroMarkTransformer;
const HtmlTransformer = require('./HtmlTransformer');

let htmlTransformer = null;
let ciceroTransformer = null;

/**
 * Prepare the text for parsing (normalizes new lines, etc)
 * @param {string} input - the text for the clause
 * @return {string} - the normalized text for the clause
 */
function normalizeNLs(input) {
    // we replace all \r and \n with \n
    let text =  input.replace(/\r/gm,'');
    return text;
}

// @ts-ignore
beforeAll(() => {
    htmlTransformer = new HtmlTransformer();
    ciceroTransformer = new CiceroMarkTransformer();
});

/**
 * Get the name and contents of all markdown test files
 * @returns {*} an array of name/contents tuples
 */
function getMarkdownFiles() {
    const result = [];
    const files = fs.readdirSync(__dirname + '/../test/data/markdown');

    files.forEach(function(file) {
        if(file.endsWith('.md')) {
            let contents = fs.readFileSync(__dirname + '/../test/data/markdown/' + file, 'utf8');
            result.push([file, contents]);
        }
    });

    return result;
}

describe('markdown <-> html', () => {
    getMarkdownFiles().forEach(([file, markdownText], i) => {
        it(`converts ${file} to html`, () => {
            const json = ciceroTransformer.fromMarkdown(markdownText, 'json');
            expect(json).toMatchSnapshot(); // (1)
            const html = htmlTransformer.toHtml(json);
            expect(html).toMatchSnapshot(); // (2)
            const ciceroMarkDom = htmlTransformer.toCiceroMark(html, 'json');
            expect(ciceroMarkDom).toEqual(json);
        });
    });

    it('converts unwrapped <li> to html', () => {
        const ciceroMarkDom = htmlTransformer.toCiceroMark('<p>Hello</p><li>list item</li><p>World.</p>', 'json');
        expect(ciceroMarkDom).toMatchSnapshot(); // (1)
        const md = ciceroTransformer.toMarkdown(ciceroMarkDom);
        expect(md).toMatchSnapshot(); // (2)
    });
});

/**
 * Get the name and contents of all ciceromark test files
 * @returns {*} an array of name/contents tuples
 */
function getCiceroMarkFiles() {
    const result = [];
    const files = fs.readdirSync(__dirname + '/../test/data/ciceromark');

    files.forEach(function(file) {
        if(file.endsWith('.json')) {
            let contents = normalizeNLs(fs.readFileSync(__dirname + '/../test/data/ciceromark/' + file, 'utf8'));
            result.push([file, contents]);
        }
    });

    return result;
}

describe('ciceromark <-> html', () => {
    getCiceroMarkFiles().forEach( ([file, jsonText], index) => {
        it(`converts ${file} to and from CiceroMark`, () => {
            const value = JSON.parse(jsonText);
            const html = htmlTransformer.toHtml(value);

            // check no changes to html
            expect(html).toMatchSnapshot(); // (1)

            // load expected html
            const expectedHtml = normalizeNLs(fs.readFileSync(__dirname + '/../test/data/ciceromark/' + file.replace(/.json$/,'.html'), 'utf8'));
            expect(expectedHtml).toMatchSnapshot(); // (2)

            // convert the expected html and compare
            const expectedCiceroMarkValue = htmlTransformer.toCiceroMark(expectedHtml);
            expect(expectedCiceroMarkValue).toMatchSnapshot(); // (3)

            // check that html created from ciceromark and from the expected html is the same
            expect(html).toEqual(expectedHtml);

            // check roundtrip
            expect(expectedCiceroMarkValue).toEqual(value);
        });
    });
});
