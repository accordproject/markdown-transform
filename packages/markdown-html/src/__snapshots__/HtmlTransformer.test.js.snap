// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html converts acceptance.md to html 1`] = `
Object {
  "$class": "org.accordproject.commonmark.Document",
  "nodes": Array [
    Object {
      "$class": "org.accordproject.commonmark.Heading",
      "level": "1",
      "nodes": Array [
        Object {
          "$class": "org.accordproject.commonmark.Text",
          "text": "HELLO! This is the contract editor.",
        },
      ],
    },
    Object {
      "$class": "org.accordproject.commonmark.Paragraph",
      "nodes": Array [
        Object {
          "$class": "org.accordproject.commonmark.Text",
          "text": "And below is a ",
        },
        Object {
          "$class": "org.accordproject.commonmark.Strong",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": "clause",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark.Text",
          "text": ".",
        },
      ],
    },
    Object {
      "$class": "org.accordproject.ciceromark.Clause",
      "clauseid": "479adbb4-dc55-4d1a-ab12-b6c5e16900c0",
      "nodes": Array [
        Object {
          "$class": "org.accordproject.commonmark.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": "Acceptance of Delivery. ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "shipper",
              "value": "\\"Party A\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": " will be deemed to have completed its delivery obligations if in ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "receiver",
              "value": "\\"Party B\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": "'s opinion, the ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "deliverable",
              "value": "\\"Widgets\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": " satisfies the Acceptance Criteria, and ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "receiver",
              "value": "\\"Party B\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": " notifies ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "shipper",
              "value": "\\"Party A\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": " in writing that it is accepting the ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "deliverable",
              "value": "\\"Widgets\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": ".",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": "Inspection and Notice. ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "receiver",
              "value": "\\"Party B\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": " will have ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "businessDays",
              "value": "10",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": " Business Days' to inspect and evaluate the ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "deliverable",
              "value": "\\"Widgets\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": " on the delivery date before notifying ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "shipper",
              "value": "\\"Party A\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": " that it is either accepting or rejecting the ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "deliverable",
              "value": "\\"Widgets\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": ".",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": "Acceptance Criteria. The \\"Acceptance Criteria\\" are the specifications the ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "deliverable",
              "value": "\\"Widgets\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": " must meet for the ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "shipper",
              "value": "\\"Party A\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": " to comply with its requirements and obligations under this agreement, detailed in ",
            },
            Object {
              "$class": "org.accordproject.ciceromark.Variable",
              "id": "attachment",
              "value": "\\"Attachment X\\"",
            },
            Object {
              "$class": "org.accordproject.commonmark.Text",
              "text": ", attached to this agreement.",
            },
          ],
        },
      ],
      "src": "ap://acceptance-of-delivery@0.12.1#721d1aa0999a5d278653e211ae2a64b75fdd8ca6fa1f34255533c942404c5c1f",
    },
  ],
  "xmlns": "http://commonmark.org/xml/1.0",
}
`;

exports[`html converts acceptance.md to html 2`] = `
"<html>
<body>
<div class=\\"document\\">
<h1>HELLO! This is the contract editor.</h1>
<p>And below is a <strong>clause</strong>.</p>
<div class=\\"clause\\" clauseid=\\"479adbb4-dc55-4d1a-ab12-b6c5e16900c0\\" src=\\"ap://acceptance-of-delivery@0.12.1#721d1aa0999a5d278653e211ae2a64b75fdd8ca6fa1f34255533c942404c5c1f\\">
<p>Acceptance of Delivery. <span class=\\"variable\\" id=\\"shipper\\">\\"Party A\\"</span> will be deemed to have completed its delivery obligations if in <span class=\\"variable\\" id=\\"receiver\\">\\"Party B\\"</span>'s opinion, the <span class=\\"variable\\" id=\\"deliverable\\">\\"Widgets\\"</span> satisfies the Acceptance Criteria, and <span class=\\"variable\\" id=\\"receiver\\">\\"Party B\\"</span> notifies <span class=\\"variable\\" id=\\"shipper\\">\\"Party A\\"</span> in writing that it is accepting the <span class=\\"variable\\" id=\\"deliverable\\">\\"Widgets\\"</span>.</p>
<p>Inspection and Notice. <span class=\\"variable\\" id=\\"receiver\\">\\"Party B\\"</span> will have <span class=\\"variable\\" id=\\"businessDays\\">10</span> Business Days' to inspect and evaluate the <span class=\\"variable\\" id=\\"deliverable\\">\\"Widgets\\"</span> on the delivery date before notifying <span class=\\"variable\\" id=\\"shipper\\">\\"Party A\\"</span> that it is either accepting or rejecting the <span class=\\"variable\\" id=\\"deliverable\\">\\"Widgets\\"</span>.</p>
<p>Acceptance Criteria. The \\"Acceptance Criteria\\" are the specifications the <span class=\\"variable\\" id=\\"deliverable\\">\\"Widgets\\"</span> must meet for the <span class=\\"variable\\" id=\\"shipper\\">\\"Party A\\"</span> to comply with its requirements and obligations under this agreement, detailed in <span class=\\"variable\\" id=\\"attachment\\">\\"Attachment X\\"</span>, attached to this agreement.</p>
</div>
</div>
</body>
</html>"
`;
